kgraphql/src/main/kotlin/com/apurebase/kgraphql/schema/structure/RequestInterpreter.kt

```diff
@@ -210,25 +210,53 @@ class RequestInterpreter(private val schemaModel: SchemaModel) {
     ): Execution.Union {
         validateUnionRequest(field, selectionNode)
 
+        fun SelectionSetNode.addSelectionNode(node: SelectionNode?) = node?.let {
+            copy(selections = selections + node)
+        } ?: this
+
+        // https://spec.graphql.org/October2021/#sec-Unions
+        //  "With interfaces and objects, only those fields defined on the type can be queried directly; to query
+        //  other fields on an interface, typed fragments must be used. This is the same as for unions, but unions
+        //  do not define any fields, so *no* fields may be queried on this type without the use of type refining
+        //  fragments or inline fragments (with the exception of the meta-field `__typename`)."
         val unionMembersChildren: Map<Type, List<Execution>> =
             field.returnType.possibleTypes.associateWith { possibleType ->
                 val selections = selectionNode.selectionSet?.selections
+                // Add __typename to all fragment selections if specified
+                val typenameSelection = selections?.firstOrNull { it is FieldNode && it.name.value == "__typename" }
 
-                val a = selections?.filterIsInstance<FragmentSpreadNode>()?.firstOrNull {
+                val namedFragment = selections?.filterIsInstance<FragmentSpreadNode>()?.firstOrNull {
```

Comment:
Related: I believe multiple fragments on the same type _should_ be allowed and then merged in the result set. Will (have to) read up on that and adjust if needed.

Reply:
Adjusted